### PR TITLE
allow for regular background updates of avatars from social networks

### PR DESCRIPTION
**Pull request by call-me-matt**
implements #1594

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Nextcloud Contacts
-![Downloads](https://img.shields.io/github/downloads/nextcloud/contacts/total.svg?style=flat-square)
+![Downloads](https://img.shields.io/github/downloads/nextcloud/contacts/total?style=flat-square)
```
Comment (skjnldsv) [inline README.md:2]: why? :thinking: :grin:

File: lib/Controller/PageController.php [hunk 2]
```diff
@@ -69,13 +71,19 @@ public function __construct(string $appName,
 	 * Default routing
 	 */
 	public function index(): TemplateResponse {
+		$userId = \OC_User::getUser();
```
Comment (skjnldsv) [inline lib/Controller/PageController.php:74]: Please inject it, do not use this :)

https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/public/IUserSession.php#L79

File: lib/Cron/SocialUpdateRegistration.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		parent::setInterval(7 * 24 * 60 * 60);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	protected function run($arguments) {
+
+		// check if admin allows for social updates:
+		$isAdminEnabled = $this->config->getAppValue($this->appName, 'allowSocialSync', 'yes');
```
Comment (skjnldsv) [inline lib/Cron/SocialUpdateRegistration.php:75]: I would go for a different variable name.
Something like `$isSyncEnabledByAdmin`

File: lib/Cron/SocialUpdateRegistration.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		if (!($isAdminEnabled === 'yes')) {
+			return;
+		}
+
+		$this->userManager->callForSeenUsers(function (IUser $user) {
+
+			// check that user opted-in:
+			$isUserEnabled = $this->config->getUserValue($user->getUID(), $this->appName, 'enableSocialSync', 'no');
```
Comment (skjnldsv) [inline lib/Cron/SocialUpdateRegistration.php:83]: same :)

File: src/components/AdminSettings.vue [hunk 5]
```diff
@@ -1,3 +1,25 @@
+<!--
+  - @copyright Copyright (c) Matthias Heinisch <nextcloud@matthiasheinisch.de>
+  -
+  - @author Matthias Heinisch <nextcloud@matthiasheinisch.de>
```
Comment (skjnldsv) [inline src/components/AdminSettings.vue:4]: :+1:

File: lib/Service/SocialApiService.php [hunk 6]
```diff
@@ -182,4 +215,82 @@ public function updateContact(string $addressbookId, string $contactId, string $
...
+	 *
+	 * Updates social profile data for all contacts of an addressbook
+	 * // TODO: how to exclude certain contacts?
+	 *
+	 * @param {String} network the social network to use (fallback: take first match)
+	 * @param {String} user the address book owner
+	 *
+	 * @returns {JSONResponse} JSONResponse with the list of changed and failed contacts
+	 */
+	public function updateAddressbooks(string $network, string $user) : JSONResponse {
```
Comment (skjnldsv) [inline lib/Service/SocialApiService.php:231]: We usually use `userId` when we use the id, $user is usually refering to the IUser class :)

File: README.md (hunk 1 above)
Comment (call-me-matt) [reply README.md:2]: No clue where that came from. I think my git has a personality.

File: src/components/AdminSettings.vue (hunk 5 above)
Comment (call-me-matt) [reply src/components/AdminSettings.vue:4]: so you know who to blame :man_juggling: :policeman:

File: lib/Service/SocialApiService.php [hunk 7]
```diff
@@ -124,6 +141,22 @@ protected function getAddressBook(string $addressbookId) : ?IAddressBook {
...
+	 * @NoAdminRequired
+	 *
+	 * Retrieves and initiates all addressbooks from a user
+	 *
+	 * @param {string} user the user to query
+	 * @param {IManager} the contact manager to load
+	 */
+	protected function registerAddressbooks($user, IManager $manager) {
+		$cm = new ContactsManager($this->davBackend, $this->l10n);
+		$cm->setupContactsProvider($manager, $user, $this->urlGen);
```
Comment (call-me-matt) [inline lib/Service/SocialApiService.php:154]: Is there a way to limit this to address books owned by a user (no system address book, no shared address books from other users)? Or, alternatively, is there a way to find out about an owner of an IAddressBook?

File: lib/Service/SocialApiService.php [hunk 8]
```diff
@@ -182,4 +215,82 @@ public function updateContact(string $addressbookId, string $contactId, string $
...
+		// TODO: keep only owned address books (not shared ones)
+
+		foreach ($addressBooks as $addressBook) {
+			if (is_null($addressBook)) {
+				break;
+			}
+
+			// get contacts in that addressbook
+			$contacts = $addressBook->search('', ['UID'], ['types' => true]);
+			// TODO: filter for contacts with social profiles
```
Comment (call-me-matt) [inline lib/Service/SocialApiService.php:265]: Could be optimized with `$addressBook->search('', ['X-SOCIALPROFILE'], ['types' => true]);`, if `X-SOCIALPROFILE` was [indexed](https://github.com/nextcloud/server/blob/master/apps/dav/lib/CardDAV/CardDavBackend.php#L78-L80)

File: lib/Service/SocialApiService.php [hunk 9]
```diff
@@ -182,4 +215,82 @@ public function updateContact(string $addressbookId, string $contactId, string $
...
+		$this->registerAddressbooks($userId, $this->manager);
+
+		$addressBooks = $this->manager->getUserAddressBooks();
+		// TODO: filter out system addr books
+		// TODO: filter out deactivated books
+		// TODO: keep only owned address books (not shared ones)
+
+		foreach ($addressBooks as $addressBook) {
+			if (is_null($addressBook)) {
+				break;
```
Comment (call-me-matt) [inline lib/Service/SocialApiService.php:257]: should be `continue`

File: lib/Service/SocialApiService.php [hunk 10]
```diff
@@ -182,4 +215,82 @@ public function updateContact(string $addressbookId, string $contactId, string $
...
+			}
+
+			// get contacts in that addressbook
+			$contacts = $addressBook->search('', ['UID'], ['types' => true]);
+			// TODO: filter for contacts with social profiles
+			if (is_null($contacts)) {
+				break;
```
Comment (call-me-matt) [inline lib/Service/SocialApiService.php:264]: should be `continue`

File: lib/Cron/SocialUpdateRegistration.php (hunk 3 above)
Comment (call-me-matt) [reply lib/Cron/SocialUpdateRegistration.php]: Now Codacy complains:

> Avoid excessively long variable names like $isSocialEnabledByAdmin. Keep variable name length under 20.

Shall I go back to the previous variable names?

File: lib/Service/SocialApiService.php (hunk 7 above)
Comment (skjnldsv) [reply lib/Service/SocialApiService.php]: cc @georgehrke what do you think? Does it make sense to add filters to the IAddressBook interface?

File: lib/Cron/SocialUpdateRegistration.php (hunk 3 above)
Comment (skjnldsv) [reply lib/Cron/SocialUpdateRegistration.php]: xD, Ignore codacy, I don't even know why I enabled it in the first place :p

File: lib/Controller/SocialApiController.php [hunk 11]
```diff
@@ -66,6 +72,38 @@ public function setAppConfig($key, $allow) {
...
+	 *
+	 * update appconfig (user setting)
+	 *
+	 * @param {String} key the identifier to change
+	 * @param {String} allow the value to set
+	 *
+	 * @returns {JSONResponse} an empty JSONResponse with respective http status code
+	 */
+	public function setUserConfig($key, $allow) {
+		$userId = $this->userSession->getUser()->getUid();
```
Comment (skjnldsv) [inline lib/Controller/SocialApiController.php:86]: Might be better to get the user and check against it being null before getting the UID :)

File: lib/Cron/SocialUpdateRegistration.php (hunk 3 above)
Comment (call-me-matt) [reply lib/Cron/SocialUpdateRegistration.php]: I think I found a compromise :wink:

File: lib/Service/SocialApiService.php (hunk 8 above)
Comment (skjnldsv) [reply lib/Service/SocialApiService.php:297]: > Could be optimized with `$addressBook->search('', ['X-SOCIALPROFILE'], ['types' => true]);`, if `X-SOCIALPROFILE` was [indexed](https://github.com/nextcloud/server/blob/master/apps/dav/lib/CardDAV/CardDavBackend.php?rgh-link-date=2020-07-31T19%3A06%3A55Z#L78-L80)

Mind to add this? :)

Comment (skjnldsv) [reply lib/Service/SocialApiService.php:297]: if it's not indexed (for older version) does it still works? Is the searched value just ignored?
If so it's ok I guess :)

Comment (call-me-matt) [reply lib/Service/SocialApiService.php]: Right, I have included this and added a fallback for older server versions. But concerning the indexer, I could not find out how to add the `X-SOCIALPROFILE`. Simply adding it to the linked array did not have any effect. Do I need to trigger the indexer? Maybe someone from the server-team can help?

Comment (call-me-matt) [reply lib/Service/SocialApiService.php]: Pull request created: https://github.com/nextcloud/server/pull/22085

Comment (call-me-matt) [reply lib/Service/SocialApiService.php]: Turns out it was not cron, but me refreshing the profiles and triggering a re-index by that. We'll need a background job to re-index when a user activates this function to make sure all contacts are found. @skjnldsv or is there a better way / place to do that?

**Comment by MorrisJobke** [comment]
Bildschirmfoto 2020-09-15 um 12 36 26

😢

**Comment by call-me-matt** [comment]
Ups, I'm sorry. That was an accident.

**Pull request merged**